dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,183 @@
+
+- Top 10 Projects with the most unique users
+- Number of unique users
+- Last download date
+- Total data size (in GiB) for each project
+- Export date (the last time this table was updated)
+"""
+
+from airflow.decorators import dag
+
+from dataclasses import dataclass
+from datetime import date, datetime
+from typing import List
+
+import synapseclient
+from airflow.decorators import dag, task
+from airflow.models.param import Param
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+
+SYNAPSE_RESULTS_TABLE = "syn61597055"
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+    "current_date": Param(date.today().strftime("%Y-%m-%d"), type="string"),
+    "time_window": Param("30", type="string"),
+    }
+
+dag_config = {
+    # run on the 2nd of every month at midnight
+    "schedule_interval": "0 0 2 * *",
+    "start_date": datetime(2024, 1, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotMetrics:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        project_id: The ID of the Project
+        n_unique_users: The number of unique users who downloaded from the Project
+        last_download_date: The date of the last download from the Project
+        total_data_size_in_gib: The current total size of the Project in GiB
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
```

Comment:
This is for code styling to make things easier to read. We can assume anything with the `AND` operator is an actual filter.

Reply:
Nice!